File: src/main/java/com/nospawnn/ui/argeditors/BoolEditor.java
package com.nospawnn.ui.argeditors;

import com.nospawnn.ui.listeners.ArgListener;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public final class BoolEditor extends ArgumentEditor {
    private JPanel contentPane;
    private JLabel label;
    private JComboBox<Boolean> comboBox;

    private final List<ArgListener> argListeners = new ArrayList<>();

    public BoolEditor(String labelText, Boolean initialValue) {
        this.label.setText(labelText);
        this.comboBox.setSelectedIndex(initialValue ? 0 : 1);
    }

    @Override
    public JPanel getContentPane() {
        return contentPane;
    }

    @Override
    public void addListener(ArgListener al) {
        argListeners.add(al);
    }

    @Override
    public void notifyListeners() {
        for (var l : argListeners) l.argUpdated(this);
    }

    @Override
    public Boolean getValue() {
        return Objects.equals(comboBox.getSelectedItem(), true);
    }

    @Override
    public JLabel getLabel() {
        return this.label;
    }

    @Override
    public String getLabelText() {
        return this.label.getText();
    }

    @Override
    public Component getEditor() {
        return this.comboBox;
    }

    void createUIComponents() {
        this.label = new JLabel();
        this.comboBox = new JComboBox<>();

        comboBox.addItem(true);
        comboBox.addItem(false);
        this.comboBox.addItemListener(e -> {
            if (e.getStateChange() == ItemEvent.DESELECTED) return;
            notifyListeners();
        });
    }

}

File: src/main/kotlin/ui/argumenteditors/BoolEditor.kt
package com.nospawnn.ui.argumenteditors

import com.nospawnn.ui.MainView
import java.awt.event.ItemEvent
import java.awt.event.ItemListener
import javax.swing.DefaultComboBoxModel
import javax.swing.JComboBox

class BoolEditor(argName: String, override val default: Boolean) : ArgumentEditor(argName) {
    override val editor = JComboBox<Boolean>().apply {
        addItem(true)
        addItem(false)
        selectedItem = if (default == true) getItemAt(0) else getItemAt(1)
    }

    init {
        editor.addItemListener(object : ItemListener {
            override fun itemStateChanged(e: ItemEvent?) {
                (topLevelAncestor as MainView).updateOutput()
            }
        })
        addEditor(editor)
    }


    override fun getValue(): Boolean = (editor.model as DefaultComboBoxModel<Boolean>).selectedItem == true
}